compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/function_outlining.mlir

```diff
@@ -0,0 +1,91 @@
+// RUN: iree-opt --split-input-file --iree-amdaie-function-outlining --verify-diagnostics --split-input-file %s | FileCheck %s
+
+// CHECK-LABEL: func.func private @generic_matmul_outlined
+// CHECK-SAME:  (%[[LHS:.*]]: memref<1x1x4x8x4x8xbf16>,
+// CHECK-SAME:   %[[RHS:.*]]: memref<1x1x8x4x8x4xbf16>,
+// CHECK-SAME:   %[[OUT:.*]]: memref<1x1x8x8x4x4xf32>) {
+// CHECK:           linalg.generic
+// CHECK-SAME:          ins(%[[LHS]], %[[RHS]] :
+// CHECK-SAME:          outs(%[[OUT]] :
+// CHECK:           return
+// CHECK:        }
+// CHECK-LABEL:  func.func @matmul_example
+// CHECK-SAME:   (%[[A:.*]]: memref<1x1x4x8x4x8xbf16>,
+// CHECK-SAME:    %[[B:.*]]: memref<1x1x8x4x8x4xbf16>,
+// CHECK-SAME:    %[[C:.*]]: memref<1x1x8x8x4x4xf32>) {
+// CHECK:            amdaie.core
+// CHECK:               func.call @generic_matmul_outlined(%[[A]], %[[B]], %[[C]])
+// CHECK-NOT:           linalg.generic
+// CHECK:               amdaie.end
+// CHECK:            }
+// CHECK:            return
+// CHECK:        }
+func.func @matmul_example(%A: memref<1x1x4x8x4x8xbf16>, %B: memref<1x1x8x4x8x4xbf16>, %C: memref<1x1x8x8x4x4xf32>) {
```

Comment:
Is the idea that if you outline the same function twice, then both calls will be to the same function? i.e. code size reduction. If so, can we have a lit test for this?  I assume that it works, because it sounds like you've got an e2e test passing now that was not passing before. But a lit test would still be useful IMO

Reply:
Hi @newling @jtuyls @yzhang93 - I have raised a separate PR for creating a standalone pass for outlining https://github.com/nod-ai/iree-amd-aie/pull/862 as this pass' creation in itself won't cause any issue - the review comments I've received for this pass has therefore been addressed and carried over to this new PR. Please take a look.

I'm keeping rest of the infra within this PR itself and would rebase this once #862 gets merged.

@newling - I have also added a lit test in this PR as rest of the infra is needed to make the e2e work. Please take a look.

Reply:
Thanks @Abhishek-Varma. I'll take a look at https://github.com/nod-ai/iree-amd-aie/pull/862 asap

Reply:
> @newling - I have also added a lit test in this PR as rest of the infra is needed to make the e2e work. Please take a look.

I was thinking a test of just the outlining pass, something like 
```
func.func @main() {
  %0 = linalg.matmul ...
  %1 = linalg.matmul ...
}
```
being transformed to 
```
func.func @outlined_matmul() {
   ...
}

func.func @main() {
   %0 = func.call @outlined_matmul ... 
   %1 = func.call @outlined_matmul ... 
}
```